Non-blocking primary relocation hand-off

ywelsch opened the pull request:
Primary relocation and indexing concurrently can currently lead to a deadlock situation as indexing operations are blocked on a (bounded) thread pool during the hand-off phase between old and new primary. This PR replaces blocking of indexing operations by putting operations that cannot be executed during relocation hand-off in a queue to be executed once relocation completes.

Relates to #18553, #15900.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -235,17 +237,48 @@ public void messageReceived(final Request request, final TransportChannel channe
-            try {
-                if (primaryShardReference.isRelocated()) {
+
+            Callback<Throwable> onFailure = t -> {
```
all these small call backs make me think we need asyncPrimaryAction which has the channel and the task as fields and implements ActionListener<PrimaryShardReference> so we can pass it as a value to the call backs?

ywelsch replied:
I would very much prefer an `AsyncPrimaryAction` (especially so that it implements AbstractRunnable, making failures simpler).

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -264,29 +297,12 @@ public void handleException(TransportException exp) {
-            }
+            };
+
+            acquirePrimaryShardReference(request.shardId(), onReferenceAcquired, onRelocated, onFailure);
```
why can't we use an ActionListener<PrimaryShardReference> call back here? I think we can deal with the relocation check in the same way did before, within the onResponse method?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -436,20 +437,27 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
-                    throw new IndexShardNotStartedException(shardId, state);
+    public void relocated(String reason) throws IllegalIndexShardStateException, InterruptedException {
+        try {
+            indexShardOperationsLock.blockOperations(30, TimeUnit.MINUTES, () -> {
```
++, one less releasable to remember to free.

bleskes replied:
although I like it, it means the implementation on the indexShardOperationsLock can not use simple synchornized methods  because their duration is unclear . if this returned a releasble like it used to, I *think* all ops over there can be synchorinze and we won't need fancy non-blocking implementations with `while(true)` . If so, I think that's the right trade off to roll this back?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java:
```diff
@@ -0,0 +1,132 @@
+                            delayedOperations = new ArrayList<>();
+                        }
+                        if (executor != null) {
+                            delayedOperations.add(new ThreadedActionListener(logger, threadPool, executor, onAcquired, true));
```
forcing execution should be a parameter for now imo - I know we want/maybe/potentially change how we deal with replicas and queues, but for now I rather not change semantics and have primary ops non-forced and replicas ops forced.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
+            try {
+                cancellableThreads.execute(() -> shard.relocated("to " + request.targetNode()));
+            } catch (Throwable e) {
+                logger.debug("[{}][{}] completing relocation hand-off to {} failed", indexName, shardId, request.targetNode(), e);
```
I think the exception is not in the right place?

ywelsch replied:
Good catch. I was surprised to see that this was not caught by the logger usage checker. After digging I noticed that the checker was accidentally disabled since it was added. I've opened #19126 to address this.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTargetService.java:
```diff
@@ -263,12 +263,6 @@ public RecoveryResponse newInstance() {
                 return;
             }
 
-            if (cause instanceof IndexShardClosedException) {
```
why did you remove this?

ywelsch replied:
It's dead code. A few lines above we match the super class
`if (cause instanceof IllegalIndexShardStateException...`

bleskes replied:
I wonder if we should move it up above the IllegalIndexShardStateException, as we don't want to retry in this case (although it's not that bad).

bleskes replied:
ping?

ywelsch replied:
I'm not sure we should change this as part of this PR (I recall some issues when I tried that, will have to check again). I can leave this dead code for now and add a TODO instead.

bleskes replied:
I'm fine with removing it - it's useless . I'm ok with just checking it further after the PR is done.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
              */
-            try {
-                shard.relocated("to " + request.targetNode());
-            } catch (IllegalIndexShardStateException e) {
```
++ on removing this catch. Not true any more

bleskes commented on core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java:
```diff
@@ -136,7 +136,7 @@ public ActionRequestValidationException validate() {
         protected void doExecute(Task task, final Request request, ActionListener<Response> listener) {
             // remove unneeded threading by wrapping listener with SAME to prevent super.doExecute from wrapping it with LISTENER
-            super.doExecute(task, request, new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.SAME, listener));
+            super.doExecute(task, request, new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.SAME, listener, false));
```
I wonder if we should make false be the default with an overload?

ywelsch replied:
It's only used in 3 places, so not worth the change imho.

bleskes replied:
k

bleskes commented:
Thx @ywelsch . I left some comments that I think will simplify things. My main concern here is the extra IndexShardOperationsLock wrapper around `SuspendableRefContainer`. I'm not sure we need an extra abstraction instead of making SuspendableRefContainer implement the API we need (or just rename it). It makes things more complex, for example, IndexShardOperationsLock assumes that the only reason why tryAcquire can fail is that a block operation is going on. This is true now, but only because we use Integer.MAX_VALUE as the total amount of operations. Some one can change that and not realize the implications it has.

ywelsch commented:
@bleskes I've updated the PR with the following main changes:
- introduced `AsyncPrimaryAction` and use the original flow of the method to distinguish `isRelocated` on the `PrimaryShardReference`.
- inlined `SuspendableRefContainer` into `IndexShardOperationsLock`.
- simplified lock acquisition retry (eliminating the loop) to remove some of the non-blocking fanciness.
Please have another look.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -157,7 +160,7 @@ protected void resolveRequest(MetaData metaData, IndexMetaData indexMetaData, Re
 
     /**
      * Synchronous replica operation on nodes with replica copies. This is done under the lock form
```
form -> from

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -746,17 +789,29 @@ protected PrimaryShardReference getPrimaryShardReference(ShardId shardId) {
         }
-        return new PrimaryShardReference(indexShard, indexShard.acquirePrimaryOperationLock());
+
+        ActionListener<Releasable> onAcquired = new ActionListener<Releasable>() {
```
it's a shame we have to wrap the call back here, but I don't see how to simplify this without doing something not nice instead, like making the PrimaryShardRefernce having a non-final releasable field...

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java:
```diff
@@ -0,0 +1,167 @@
+                        delayedOperations = new ArrayList<>();
+                    }
+                    if (executor != null) {
+                        delayedOperations.add(new ThreadedActionListener(logger, threadPool, executor, onAcquired, forceExecution));
```
should be ThreadedActionListener<>

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java:
```diff
@@ -0,0 +1,167 @@
+     * @throws InterruptedException if calling thread is interrupted
+     * @throws TimeoutException if timed out waiting for in-flight operations to finish
+     */
+    public void blockOperations(long timeout, TimeUnit timeUnit, Runnable onBlocked) throws InterruptedException, TimeoutException {
```
I this this can simplified even further : https://gist.github.com/bleskes/0bf520c969eaa9542b1deefb4a8e1d5a  (also note the test fixes, which are unrelated)

ywelsch replied:
ok, I was afraid of lock contention by adding synchronization to the acquire method. We can try it your way and see how it impacts benchmarks.

bleskes replied:
If you prefer, in the indexing method we can do tryAcquire first out of lock and go under lock and try again, if failed.

ywelsch replied:
Let's try it the simple way first (tryAcquire under lock) and see if it has an impact on indexing performance. Just FYI, I've adapted your implementation of `acquire` as it was holding onto the mutex when calling `onAcquired.onResponse(releasable)`.

bleskes commented on core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java:
```diff
@@ -360,22 +359,21 @@ public void finalizeRecovery() {
         cancellableThreads.execute(recoveryTarget::finalizeRecovery);
 
         if (isPrimaryRelocation()) {
+            cancellableThreads.checkForCancel();
```
executes already checks for cancel. No need.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -629,26 +629,15 @@ public void execute() throws Exception {
-        };
-        try {
-            primaryPhase.messageReceived(request, createTransportChannel(listener), task);
-        } catch (ElasticsearchException e) {
```
awesome

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Before
+    public void createIndexShardOperationsLock() {
+         block = new IndexShardOperationsLock(logger, threadPool);
```
I wonder if we want to add an `@After` rule that checks that all semaphore permits are back.

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+        for (PlainActionFuture<Releasable> future : futures) {
+            Releasable releasable = future.get(1, TimeUnit.MINUTES);
+            assertNotNull(releasable);
+            releasable.close();
```
this causes a deadlock when the operations that are blocked by the block are first in the future list while the last part of the future list prevents the block from being acquired. See the gist I made.

ywelsch replied:
right

bleskes replied:
I think this is still a problem, no?

ywelsch replied:
uh, this change must haven been lost, will fix it.

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+
+    public void testOperationsInvokedImmediatelyIfNoBlock() throws ExecutionException, InterruptedException {
+        PlainActionFuture<Releasable> future = new PlainActionFuture<>();
+        block.acquire(future, ThreadPool.Names.GENERIC, true);
```
maybe pass an executor that doesn't exist and will blow up?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1534,29 +1542,36 @@ private final EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
+     * ActionListener will be called on the calling thread. During relocation hand-off, lock acquisition can be delayed. The provided
+     * ActionListener will then be called using the provided executor.
+     */
+    public void acquirePrimaryOperationLock(ActionListener<Releasable> onLockAcquired, String executor) {
```
maybe call the executor , executorOnDelay?

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java:
```diff
@@ -0,0 +1,167 @@
+     * @param executor executor to use for delayed call
+     * @param forceExecution whether the runnable should force its execution in case it gets rejected
+     */
+    public void acquire(ActionListener<Releasable> onAcquired, String executor, boolean forceExecution) {
```
nit: use executorOnDelay as a name?

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+
+    protected Thread blockAndWait(final CountDownLatch releaseBlock) throws InterruptedException {
+        CountDownLatch blockAcquired = new CountDownLatch(1);
+        Thread blockingThread = new Thread() {
```
just use the thread pool?

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+        CountDownLatch releaseBlock = new CountDownLatch(1);
+
+        final IndexShardClosedException exception = new IndexShardClosedException(new ShardId("blubb", "id", 0));
+        Thread blockingThread = new Thread() {
```
use thread pool?

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,251 @@
+        PlainActionFuture<Releasable> future1 = new PlainActionFuture<>();
+        block.acquire(future1, ThreadPool.Names.GENERIC, true);
+        assertTrue(future1.isDone());
+        assertThat(block.getActiveOperationsCount(), equalTo(1));
```
I'm confused by this - how is the active operation 1?

bleskes replied:
Oh. You need to explicitly close the op.

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -898,12 +918,47 @@ public void testLockingBeforeAndAfterRelocated() throws Exception {
+        });
+
+        recoveryThread.start();
+        latch.countDown();
```
it seems that the latch is useless? or maybe you wanted to enable the recovery half way through the "indexing operations"?  Also, this suffers from the same racing condition - if the ops that are now in flight when the relocation comes in are not first in the onLockAcquiredActions list, we will have a deadlock.

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -436,20 +437,27 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
+        try {
+            indexShardOperationsLock.blockOperations(30, TimeUnit.MINUTES, () -> {
+                // no shard operation locks are being held here, move state from started to relocated
+                synchronized (mutex) {
```
can we assert that there are no ops in flight?

bleskes commented on core/src/test/java/org/elasticsearch/recovery/RelocationIT.java:
```diff
@@ -433,22 +433,23 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
-            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
-            .put(indexSettings())));
+            .put(indexSettings())
+            .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 0) // TODO: set this to randomInt(halfNodes - 1) once replica data loss is fixed
```
can we open an issue for this and link to it with norelease so we won't forget?

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java:
```diff
@@ -1080,7 +1080,11 @@ private void assertShardIndexCounter() {
                 for (IndexShard indexShard : indexService) {
-                    assertThat("index shard counter on shard " + indexShard.shardId() + " on node " + nodeAndClient.name + " not 0", indexShard.getActiveOperationsCount(), equalTo(0));
+                    try {
+                        assertBusy(() -> assertThat("index shard counter on shard " + indexShard.shardId() + " on node " + nodeAndClient.name + " not 0", indexShard.getActiveOperationsCount(), equalTo(0)));
```
why is this needed?

ywelsch replied:
The assertion trips on my CI when I don't put assertBusy() there. I'll have to investigate why.

ywelsch replied:
ok, I figured this one out. We have to release the shard operation lock before we respond to the caller, otherwise we might still hold on to the lock (for a very short time) after tests are finished.

bleskes commented:
@ywelsch I liked it! left a bunch of comments.

ywelsch commented:
@bleskes I've pushed a new set of changes addressing your comments. I've also added a `close` method to IndexShardOperationLock. This is not strictly necessary but makes operations fail faster when the shard is closed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -235,12 +238,33 @@ public void messageReceived(final Request request, final TransportChannel channe
+        public void onResponse(PrimaryShardReference primaryShardReference) {
             try {
                 if (primaryShardReference.isRelocated()) {
+                    primaryShardReference.close(); // release shard operation lock as soon as possible
```
do we have a test that test that double closing on the shard op lock is OK?

ywelsch replied:
yes, here: https://github.com/elastic/elasticsearch/pull/19013/files/a8a187b6e7f04c1621b1fa1db4f29f8832b91fc6#diff-322f042a012dbd1c09a9b43f40636655R180

bleskes replied:
++

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardOperationsLockTests.java:
```diff
@@ -0,0 +1,196 @@
+            thread.start();
+        }
+
+        threadPool.generic().execute(() -> {
```
I think this can move to before the threads start -  will be nastier

ywelsch replied:
ok, done in ae11e7e

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShardOperationsLock.java:
```diff
@@ -0,0 +1,166 @@
+     * @param forceExecution whether the runnable should force its execution in case it gets rejected
+     */
+    public void acquire(ActionListener<Releasable> onAcquired, String executorOnDelay, boolean forceExecution) {
+        if (closed) {
```
can we add a little test for the is new behavior?

ywelsch replied:
done in ae11e7e

bleskes commented:
LGTM. Thx @ywelsch
